snapshots/manager.go

```diff
@@ -119,11 +152,41 @@ func (m *Manager) Create(height uint64) (*types.Snapshot, error) {
 			"a more recent snapshot already exists at height %v", latest.Height)
 	}
 
-	chunks, err := m.target.Snapshot(height, types.CurrentFormat)
-	if err != nil {
-		return nil, err
-	}
-	return m.store.Save(height, types.CurrentFormat, chunks)
+	// Spawn goroutine to generate snapshot chunks and pass their io.ReadClosers through a channel
+	ch := make(chan io.ReadCloser)
+	go func() {
+		streamWriter := NewStreamWriter(ch)
+		if streamWriter == nil {
+			return
+		}
+		defer streamWriter.Close()
+		if err = m.multistore.Snapshot(height, streamWriter); err != nil {
+			streamWriter.CloseWithError(err)
+			return
+		}
+		for _, name := range m.sortedExtensionNames() {
+			extension := m.extensions[name]
+			// write extension metadata
+			err := streamWriter.WriteMsg(&types.SnapshotItem{
+				Item: &types.SnapshotItem_Extension{
+					Extension: &types.SnapshotExtensionMeta{
+						Name:   name,
+						Format: extension.SnapshotFormat(),
+					},
+				},
+			})
+			if err != nil {
+				streamWriter.CloseWithError(err)
+				return
+			}
+			if err := extension.Snapshot(height, streamWriter); err != nil {
+				streamWriter.CloseWithError(err)
+				return
+			}
+		}
+	}()
```

Comment:
Shouldn't we run the whole `Create` in a go routine rather than extensions only?
There might be some "yellow flag" issue: in https://github.com/cosmos/cosmos-sdk/pull/10961/files#diff-84a22ba89cc7868bbafc59e07c35ba0bffb05115962cb885bf39eef1a1a8a66eR144  we call end, which will happen before this clousure will finish.

Reply:
- The `BaseApp.Commit` will run the whole snapshot thing in a goroutine, but we still need the goroutine here to avoid deadlock on the channel, and the channel is nesserary for the reader/writer interfaces.
- sorry I can't pinpoint the code referenced by the link.